File: backend/app/utils/streaming.py
import json
from collections.abc import AsyncGenerator

import aiohttp

from app.config import settings


async def stream_chat_completion(
    request_body: dict,
) -> AsyncGenerator[str, None]:
    api_url = settings.OPENAI_API_BASE_URL
    api_key = settings.OPENAI_API_KEY

    headers = {
        "Authorization": f"Bearer {api_key}",
        "Content-Type": "application/json",
    }

    body = {**request_body, "stream": True}

    async with aiohttp.ClientSession() as session:
        async with session.post(
            f"{api_url}/chat/completions",
            headers=headers,
            json=body,
        ) as response:
            if response.status != 200:
                error_text = await response.text()
                error_data = {
                    "error": {
                        "message": f"LLM API error ({response.status}): {error_text}",
                        "type": "upstream_error",
                    }
                }
                yield f"data: {json.dumps(error_data)}\n\n"
                yield "data: [DONE]\n\n"
                return

            async for line in response.content:
                decoded = line.decode("utf-8").strip()
                if decoded.startswith("data: "):
                    yield decoded + "\n\n"
                    if decoded == "data: [DONE]":
                        return

File: frontend/src/lib/apis/streaming.ts
import { getToken } from '$lib/utils';
import type { ChatCompletionRequest } from '$lib/types';

export async function* streamChatCompletion(
	body: ChatCompletionRequest,
	chatId?: string
): AsyncGenerator<string> {
	const token = getToken();
	const url = chatId
		? `/api/chat/completions?chat_id=${chatId}`
		: '/api/chat/completions';

	const response = await fetch(url, {
		method: 'POST',
		headers: {
			'Content-Type': 'application/json',
			...(token ? { Authorization: `Bearer ${token}` } : {})
		},
		body: JSON.stringify({ ...body, stream: true })
	});

	if (!response.ok) {
		const error = await response.json().catch(() => ({ detail: 'Stream error' }));
		throw new Error(error.detail || 'Stream error');
	}

	const reader = response.body!.getReader();
	const decoder = new TextDecoder();
	let buffer = '';

	while (true) {
		const { done, value } = await reader.read();
		if (done) break;

		buffer += decoder.decode(value, { stream: true });
		const lines = buffer.split('\n');
		buffer = lines.pop() || '';

		for (const line of lines) {
			const trimmed = line.trim();
			if (trimmed.startsWith('data: ') && trimmed !== 'data: [DONE]') {
				try {
					const json = JSON.parse(trimmed.slice(6));
					const content = json.choices?.[0]?.delta?.content;
					if (content) yield content;
				} catch {
					// skip malformed JSON
				}
			}
		}
	}
}
